tests/test_helper.hpp

```diff
@@ -0,0 +1,193 @@
+#include <cmath>
+#include <iomanip>
+#include <sycl/sycl.hpp>
+
+#include "sycl_ext_complex.hpp"
+
+#define SYCL_CPLX_TOL_ULP 5
+
+#define __half_epsilon 9.765625e-04f
+#define __half_min 6.103515625e-05f
+
+// Helpers for displaying results
+
+template <typename T> const char *get_typename() { return "Unknown type"; }
+template <> const char *get_typename<double>() { return "double"; }
+template <> const char *get_typename<float>() { return "float"; }
+template <> const char *get_typename<sycl::half>() { return "sycl::half"; }
+
+// Helper for testing each decimal type
+
+template <template <typename> typename action, typename... argsT>
+bool test_valid_types(argsT... args) {
+  bool test_passes = true;
+
+  {
+    action<double> test;
+    test_passes &= test(args...);
+  }
+
+  {
+    action<float> test;
+    test_passes &= test(args...);
+  }
+
+  {
+    action<sycl::half> test;
+    test_passes &= test(args...);
+  }
+
+  return test_passes;
+}
+
+// Helpers for comparison
+
+bool almost_equal(double x, double y, int ulp) {
+  if (std::isnan(x) && std::isnan(y))
+    return true;
+  else if (std::isinf(x) && std::isinf(y))
+    return true;
+
+  return std::abs(x - y) <=
+             std::numeric_limits<double>::epsilon() * std::abs(x + y) * ulp ||
+         std::abs(x - y) < std::numeric_limits<double>::min();
+}
+
+bool almost_equal(float x, float y, int ulp) {
+  if (std::isnan(x) && std::isnan(y))
+    return true;
+  else if (std::isinf(x) && std::isinf(y))
+    return true;
+
+  return std::abs(x - y) <=
+             std::numeric_limits<float>::epsilon() * std::abs(x + y) * ulp ||
+         std::abs(x - y) < std::numeric_limits<float>::min();
+}
+
+bool almost_equal(sycl::half x, sycl::half y, int ulp) {
+  if (sycl::isnan(x) && sycl::isnan(y))
+    return true;
+  else if (sycl::isinf(x) && sycl::isinf(y))
+    return true;
+
+  return std::abs(x - y) <= __half_epsilon * std::abs(x + y) * ulp ||
+         std::abs(x - y) < __half_min;
+}
+
+bool almost_equal(sycl::ext::cplx::complex<double> x,
```

Comment:
Do we need all those specializations?

Reply:
I had them because `almost_equal` was being used for both scalar and complex values so templating it would be tricky. I have split it up into `almost_equal_scalar` and `almost_equal_cplx` and removed the specializations.